### PR TITLE
Added QuizVC functionality with mock data

### PR DESCRIPTION
**Pull request by marinmarkelic**
https://user-images.githubusercontent.com/101728747/179504788-a69cf3ba-b612-4d7c-9349-b5f607b9f012.mp4

### CODE REVIEW

File: QuizApp/QuizApp/Mock/MockData.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import UIKit
+
+public let sportColor = UIColor(red: 242 / 255, green: 201 / 255, blue: 76 / 255, alpha: 1)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Mock/MockData.swift]: Please create a new file named Colors and add:
```
extension UIColor {
    static let sportCategory = UIColor(red: 242 / 255, green: 201 / 255, blue: 76 / 255, alpha: 1)
    ...
}
```

and use the constants in the app.

File: QuizApp/QuizApp/Quiz/CategoryCell.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import UIKit
+
+class CategoryCell: UICollectionViewCell {
+
+    static let reuseIdentifier = String(describing: CategorySlider.self)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategoryCell.swift]: `static let reuseIdentifier = String(describing: CategoryCell.self)`

File: QuizApp/QuizApp/Quiz/CategoryCell.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func set(category: Category) {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategoryCell.swift:21]: simplify this as discussed

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func reloadWith(categories: [Category]) {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: `func reload(with categories: [Category])`

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        collectionView.reloadData()
+    }
+
+}
+
+extension CategorySlider: ConstructViewsProtocol {
+
+    func createViews() {
+        collectionViewLayout = UICollectionViewFlowLayout()
+
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: remove this line

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    func styleViews() {
+        collectionViewLayout.scrollDirection = .horizontal
+
+        collectionView.collectionViewLayout = collectionViewLayout
+        collectionView.register(CategoryCell.self, forCellWithReuseIdentifier: CategoryCell.reuseIdentifier)
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        collectionView.backgroundColor = .none
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: check if this is needed

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    ) -> CGSize {
+
+        let text = categories[indexPath.row].name
+        let font = UIFont(name: "SourceSansPro-Bold", size: 20) ?? UIFont.systemFont(ofSize: 20)
+
+        let size = text.size(withAttributes: [
+            .font: font
+        ])
+
+        let itemWidth = CGFloat(size.width + 5)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: do not use magic numbers, store them in named property

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        cellForItemAt indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(
+            withReuseIdentifier: CategoryCell.reuseIdentifier,
+            for: indexPath) as? CategoryCell
+        else { fatalError() }
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: initialize new cell

File: QuizApp/QuizApp/User defined types/Quiz.swift [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+case medium
+case hard
+
+}
+
+public struct Category {
+
+    let name: String
+    let color: UIColor
+    let isSelected: Bool
```
Comment (istaudacher) [inline QuizApp/QuizApp/User defined types/Quiz.swift]: you do not need this, set color in VM

File: QuizApp/QuizApp/Quiz/QuizViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import Combine
+
+class QuizViewModel {
+
+    @Published var quizes: [Quiz] = []
+    @Published var categories: [Category] = []
+
+    func change(category: Category) {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizViewModel.swift:8]: Modify this logic as discussed

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/QuizApp/Quiz/CategorySlider.swift (hunk 6 above)
Comment (marinmarkelic) [reply QuizApp/QuizApp/Quiz/CategorySlider.swift]: It is

File: QuizApp/QuizApp/Quiz/CategoryCell.swift [hunk 11]
```diff
@@ -0,0 +1,51 @@
+import UIKit
+
+class CategoryCell: UICollectionViewCell {
+
+    static let reuseIdentifier = String(describing: CategoryCell.self)
+
+    var category: Category?
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategoryCell.swift]: Why do you need this?

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+}
+
+extension CategorySlider: UICollectionViewDelegate {
+
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        guard let cell = collectionView.cellForItem(at: indexPath) as? CategoryCell else { return }
+
+        guard let category = cell.category else { return }
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: Do not fetch category data from the cell, you have `private var categories: [Category] = []`, use that.

File: QuizApp/QuizApp/Quiz/DifficultyView.swift [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+        square2 = UIImageView(image: square)
+        addSubview(square2)
+
+        square3 = UIImageView(image: square)
+        addSubview(square3)
+    }
+
+    func styleViews() {
+        square1.transform = CGAffineTransform(rotationAngle: 0.79)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/DifficultyView.swift]: Nice angle :D

File: QuizApp/QuizApp/Quiz/DifficultyView.swift [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        square2.transform = CGAffineTransform(rotationAngle: 0.79)
+        square3.transform = CGAffineTransform(rotationAngle: 0.79)
+    }
+
+    func defineLayoutForViews() {
+        square1.snp.makeConstraints {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/DifficultyView.swift:65]: It is better to contrain views next to each other. If this view becomes too small, they will overlap.

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 15]
```diff
@@ -0,0 +1,101 @@
+import UIKit
+
+class QuizCell: UICollectionViewCell {
+
+    static let reuseIdentifier = String(describing: QuizView.self)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: `static let reuseIdentifier = String(describing: QuizCell.self)`

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+    private var image: UIImageView!
+
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: remove this

File: QuizApp/QuizApp/Quiz/QuizViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    private func getCategory(for name: CategoryName) -> Category {
+        for index in (0..<categories.count)
+        where categories[index].name == name {
+            return categories[index]
+        }
+
+        return(Category(name: name))
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizViewModel.swift]: `return Category(name: name)`

File: QuizApp/QuizApp/Quiz/QuizViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        changeQuiz(category: &newCategory)
+
+        for index in (0..<categories.count) {
+            if categories[index].name == category.name {
+                categories[index] = newCategory
+            }
+        }
+    }
+
+    private func changeQuiz(category: inout Category) {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizViewModel.swift]: Do not modify existing struct, create a new one.

File: QuizApp/QuizApp/User defined types/Quiz.swift [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    case politics = "Politics"
+    case youtube = "Youtube"
+    case animals = "Animals"
+
+}
+
+public struct Category: Equatable {
+
+    let name: CategoryName
+    var color: UIColor = .white
```
Comment (istaudacher) [inline QuizApp/QuizApp/User defined types/Quiz.swift]: let

**Review by istaudacher — CHANGES_REQUESTED** [review]

File: QuizApp/QuizApp/Quiz/CategoryCell.swift [hunk 20]
```diff
@@ -0,0 +1,48 @@
+import UIKit
+
+class CategoryCell: UICollectionViewCell {
+
+    static let reuseIdentifier = String(describing: CategoryCell.self)
+
+    private var label: UILabel!
+
+    private var color: UIColor!
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategoryCell.swift]: remove this

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 21]
```diff
@@ -0,0 +1,127 @@
+import UIKit
+
+class CategorySlider: UIView {
+
+    private var categories: [Category] = []
+
+    private var collectionView: UICollectionView!
+    private var collectionViewLayout: UICollectionViewFlowLayout!
+
+    var delegate: CategorySliderDelegate!
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: weak var

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func reload(with categories: [Category]) {
+        self.categories = categories
+
+        collectionView.reloadData()
+    }
+
+    func reloadCells() {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: redraw()

File: QuizApp/QuizApp/Quiz/CategorySlider.swift (hunk 6 above)
Comment (istaudacher) [reply QuizApp/QuizApp/Quiz/CategorySlider.swift]: Can you try clear?

File: QuizApp/QuizApp/Quiz/CategorySlider.swift [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+extension CategorySlider: UICollectionViewDelegateFlowLayout {
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        layout collectionViewLayout: UICollectionViewLayout,
+        sizeForItemAt indexPath: IndexPath
+    ) -> CGSize {
+        let widthOffset = CGFloat(5)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/CategorySlider.swift]: `let widthOffset: CGFloat = 5`

File: QuizApp/QuizApp/Quiz/DifficultyView.swift [hunk 24]
```diff
@@ -0,0 +1,89 @@
+import UIKit
+
+class DifficultyView: UIView {
+
+    private var difficultyColor: UIColor!
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/DifficultyView.swift]: remove this

File: QuizApp/QuizApp/Quiz/DifficultyView.swift [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+            $0.leading.equalToSuperview()
+            $0.centerY.equalToSuperview()
+        }
+
+        square2.snp.makeConstraints {
+            $0.leading.equalTo(square1.snp.trailing).offset(5)
+            $0.centerY.equalToSuperview()
+        }
+
+        square3.snp.makeConstraints {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/DifficultyView.swift:74]: add trailing

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 26]
```diff
@@ -0,0 +1,100 @@
+import UIKit
+
+class QuizCell: UICollectionViewCell {
+
+    static let reuseIdentifier = String(describing: QuizCell.self)
+
+    private var image: UIImageView!
+    private var difficultyView: DifficultyView!
+
+    private var color: UIColor!
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: remove these

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+    private var title: UILabel!
+    private var desc: UILabel!
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+
+        backgroundColor = .red
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: remove this

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    func set(quiz: Quiz) {
+        title.text = quiz.name
+        desc.text = quiz.description
+
+        difficultyView.set(difficulty: quiz.difficulty, color: quiz.category.color)
+    }
+
+    func set(color: UIColor) {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: remove

File: QuizApp/QuizApp/Quiz/QuizCell.swift [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+        layer.masksToBounds = true
+        backgroundColor = .white.withAlphaComponent(0.3)
+
+        title.textColor = .white
+        title.font = UIFont(name: "SourceSansPro-Bold", size: 24)
+
+        desc.textColor = .white
+        desc.font = UIFont(name: "SourceSansPro-Bold", size: 14)
+        desc.numberOfLines = 0
+
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizCell.swift]: remove

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 30]
```diff
@@ -0,0 +1,98 @@
+import UIKit
+
+class QuizView: UIView {
+
+    private var quizes: [Quiz] = []
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: quizzes

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+        createViews()
+        styleViews()
+        defineLayoutForViews()
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func reloadCells() {
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: redraw

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+        collectionViewLayout.invalidateLayout()
+    }
+
+}
+
+extension QuizView: ConstructViewsProtocol {
+
+    func createViews() {
+        collectionViewLayout = UICollectionViewFlowLayout()
+
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: remove

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+        addSubview(collectionView)
+    }
+
+    func styleViews() {
+        collectionView.collectionViewLayout = collectionViewLayout
+        collectionView.register(QuizCell.self, forCellWithReuseIdentifier: QuizCell.reuseIdentifier)
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        collectionView.backgroundColor = .none
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: try clear

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        layout collectionViewLayout: UICollectionViewLayout,
+        sizeForItemAt indexPath: IndexPath
+    ) -> CGSize {
+        let itemWidth = CGFloat(frame.width)
+        let itemHeight = CGFloat(140)
+
+        return CGSize(width: itemWidth, height: itemHeight)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: let itemHeight: CGFloat = 140

File: QuizApp/QuizApp/Quiz/QuizView.swift [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+extension QuizView: UICollectionViewDelegateFlowLayout {
+
+    func collectionView(
+        _ collectionView: UICollectionView,
+        layout collectionViewLayout: UICollectionViewLayout,
+        sizeForItemAt indexPath: IndexPath
+    ) -> CGSize {
+        let itemWidth = CGFloat(frame.width)
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizView.swift]: You do not need this

File: QuizApp/QuizApp/Quiz/QuizViewModel.swift [hunk 36]
```diff
@@ -0,0 +1,86 @@
+import Combine
+import UIKit
+
+class QuizViewModel {
+
+    @Published var quizes: [Quiz] = []
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizViewModel.swift]: quizzes

File: QuizApp/QuizApp/User defined types/Quiz.swift [hunk 37]
```diff
@@ -0,0 +1,34 @@
...
+
+enum Difficulty {
+
+    case easy
+    case medium
+    case hard
+
+}
+
+enum CategoryName: String {
```
Comment (istaudacher) [inline QuizApp/QuizApp/User defined types/Quiz.swift]: CategoryType
remove : String, use : CaseIterable

File: QuizApp/QuizApp/User defined types/Quiz.swift [hunk 38]
```diff
@@ -0,0 +1,34 @@
...
+    case sport = "Sport"
+    case politics = "Politics"
+    case youtube = "Youtube"
+    case animals = "Animals"
+
+}
+
+public struct Category: Equatable {
+
+    let name: CategoryName
```
Comment (istaudacher) [inline QuizApp/QuizApp/User defined types/Quiz.swift]: let type: CategoryType

add to this class
```
var name: String {
    switch type {
        case .sport:
            return "Sport"
       ....
    }

```

File: QuizApp/QuizApp/Quiz/QuizViewModel.swift [hunk 39]
```diff
@@ -0,0 +1,86 @@
...
+    @Published var categories: [Category] = []
+
+    func change(category: Category) {
+        resetCategoriesColors()
```
Comment (istaudacher) [inline QuizApp/QuizApp/Quiz/QuizViewModel.swift]:
```
categories = CategoryType.allCases.map {
    Category(type: $0, color: $0 == category.type ? findColor($0) : .white)
}
```

**Review by istaudacher — CHANGES_REQUESTED** [review]

**Review by istaudacher — APPROVED** [review]

**Pull request merged**